include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,250 @@
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_trilinos_nox
+#define dealii_trilinos_nox
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_TRILINOS
+
+#  include <deal.II/base/exceptions.h>
+
+#  include <deal.II/lac/solver_control.h>
+
+#  include <Teuchos_ParameterList.hpp>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace TrilinosWrappers
+{
+  // Indicate that NOXSolver has not converged.
+  DeclException0(ExcNOXNoConvergence);
+
+
+  /**
+   * Wrapper around the non-linear solver from the NOX
+   * packge (https://docs.trilinos.org/dev/packages/nox/doc/html/index.html),
+   * targeting deal.II data structures.
+   */
+  template <typename VectorType>
+  class NOXSolver
+  {
+  public:
+    /**
+     * Struct that helps to configure NOXSolver. More advanced
+     * parameters are passed to the constructor NOXSolver
+     * directly via a Teuchos::ParameterList.
+     */
+    struct AdditionalData
+    {
+    public:
+      /**
+       * Constructor.
+       */
+      AdditionalData(const unsigned int max_iter                       = 10,
+                     const double       abs_tol                        = 1.e-20,
+                     const double       rel_tol                        = 1.e-5,
+                     const unsigned int threshold_nonlinear_iterations = 1,
+                     const unsigned int threshold_n_linear_iterations  = 0,
+                     const bool         reuse_solver                   = false);
+
+      /**
+       * Max number of non-linear iterations.
+       */
+      unsigned int max_iter;
+
+      /**
+       * Absolute l2 tolerance to be reached.
+       */
+      double abs_tol;
+
+      /**
+       * Relative l2 tolerance to be reached.
+       */
+      double rel_tol;
+
+      /**
+       * Number of non-linear iterations after which the preconditioner
+       * should be updated.
+       */
+      unsigned int threshold_nonlinear_iterations;
+
+      /**
+       * Max number of linear iterations after which the preconditioner
+       * should be updated. This is only used if a lambda is attached to
+       * solve_with_jacobian_and_track_n_linear_iterations.
+       */
+      unsigned int threshold_n_linear_iterations;
+
+      /**
+       * Reuse NOX solver instance in the next non-linear solution.
+       */
+      bool reuse_solver;
+    };
+
+    /**
+     * Constructor.
+     *
+     * If @p parameters is not filled, a Newton solver with full step is used.
+     * An overview of possible parameters is given at
+     * https://docs.trilinos.org/dev/packages/nox/doc/html/parameters.html.
+     */
+    NOXSolver(AdditionalData &                            additional_data,
+              const Teuchos::RCP<Teuchos::ParameterList> &parameters =
+                Teuchos::rcp(new Teuchos::ParameterList));
+
+    /**
+     * Clear internal state.
+     */
+    void
+    clear();
+
+    /**
+     * Solve non-linear problem and return number of iterations.
```

Comment:
```suggestion
     * Solve the non-linear problem and return the number of non-linear iterations executed.
```

(Iterations could somehow refer to the linear solver iterations, so some disambiguation is required.)